blockchain/difficulty.go

```diff
@@ -254,11 +278,90 @@ func (b *BlockChain) calcNextRequiredDifficulty(lastNode *blockNode, newBlockTim
 		return lastNode.bits, nil
 	}
 
-	// If we're still using a legacy algorithm
-	if algorithm != DifficultyDAA {
+	switch algorithm {
+	case DifficultyLegacy, DifficultyEDA:
 		return b.calcLegacyRequiredDifficulty(lastNode, newBlockTime, algorithm)
+	case DifficultyDAA:
+		return b.calcDAARequiredDifficulty(lastNode, newBlockTime)
+	case DifficultyAsert:
+		if referenceNode == nil {
+			// If block 1 has a median timestamp less than the activation time and..
+			// block 2 has a median timestamp greater than or equal to the activation time,
+			// then block 3 is the first block to contain the new rules and block 2 is the
+			// "reference block".
+			node := b.bestChain.Tip()
+			for {
+				if uint64(node.CalcPastMedianTime().Unix()) < b.chainParams.AxionActivationTime {
+					referenceNode = b.bestChain.next(node)
+					break
+				}
+				node = node.parent
+			}
+		}
+		return b.calcAsertRequiredDifficulty(lastNode, referenceNode.height, referenceNode.parent.timestamp, referenceNode.bits, lastNode.height+1, newBlockTime.Unix())
 	}
+	return 0, errors.New("unknown difficulty algorithm")
+}
+
+func (b *BlockChain) calcAsertRequiredDifficulty(lastNode *blockNode, referenceBlockHeight int32, anchorBlockTime int64, referenceBlockBits uint32, evalBlockHeight int32, evalBlockTime int64) (uint32, error) {
+	// For networks that support it, allow special reduction of the
+	// required difficulty once too much time has elapsed without
+	// mining a block.
+	if b.chainParams.ReduceMinDifficulty {
+		// Return minimum difficulty when more than the desired
+		// amount of time has elapsed without mining a block.
+		reductionTime := int64(b.chainParams.MinDiffReductionTime /
+			time.Second)
+		allowMinTime := lastNode.timestamp + reductionTime
+		if evalBlockTime > allowMinTime {
+			return b.chainParams.PowLimitBits, nil
+		}
+	}
+
+	target := CompactToBig(referenceBlockBits)
+
+	tDelta := evalBlockTime - anchorBlockTime
+	hDelta := evalBlockHeight - referenceBlockHeight
+
+	exponent := ((tDelta - int64(idealBlockTime)*int64(hDelta+1)) * radix) / tau
```

Comment:
I don't _think_ this can overflow the int64. It would require a timestamp very far in the future right?

Should we consider making this a big int or is it fine as is?

Reply:
Needs to be consistent with the other implementations. I would have gone with bigint, but consistency is key here. =)